rs/sns/governance/src/extensions.rs

```diff
@@ -118,21 +123,76 @@ impl ExtensionOperationSpec {
         &self.description
     }
 
-    pub fn validate(&self, arg: ExtensionOperationArg) -> Result<ValidatedOperationArg, String> {
-        (self.validate)(arg)
+    async fn validate_operation_arg(
+        &self,
+        governance: &Governance,
+        arg: ExtensionOperationArg,
+    ) -> Result<ValidatedOperationArg, String> {
+        (self.validate)(governance, arg).await
     }
 }
 
 /// Validates deposit operation arguments
-fn validate_deposit_operation(arg: ExtensionOperationArg) -> Result<ValidatedOperationArg, String> {
-    ValidatedDepositOperationArg::try_from(arg).map(ValidatedOperationArg::TreasuryManagerDeposit)
+fn validate_deposit_operation(
+    governance: &Governance,
+    arg: ExtensionOperationArg,
+) -> BoxFuture<Result<ValidatedOperationArg, String>> {
+    Box::pin(async move {
+        let structually_valid = ValidatedDepositOperationArg::try_from(arg)?;
+
+        let sns_subaccount = governance.sns_treasury_subaccount();
+        let icp_subaccount = governance.icp_treasury_subaccount();
+
+        // Fail if either is asking for more than 50% of current balance.  The balance could have changed
+        // since the proposal was created, and we don't assume that the proposal should work
+        let sns_balance = governance
```

Comment:
Nit

Since we do the exact same thing with both tokens independently, the common part could be factored out.

e.g.

```
governance.within_50_per_cent_of_treasury(sns_requested, sns_subaccount).await?;
governance.within_50_per_cent_of_treasury(icp_requested, icp_subaccount).await?;
```

Reply:
Might make sense eventually, but it is slightly more complex than that, and I'm not sure there'd be a nice shared interface between them since they use two different ledgers.   If we end up doing this somewhere else, we should probably find a nice way to deduplicate.